analysis/graph_helper.cpp

```diff
@@ -0,0 +1,127 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <unordered_map>
+#include <vector>
+#include <queue>
+#include <set>
+#include <string>
+#include <tuple>
+#include <sstream>
+#include <omp.h>
+
+namespace py = pybind11;
+
+class Graph {
+public:
+    std::unordered_map<unsigned int, std::unordered_set<unsigned int>> graph;
+
+    std::vector<std::vector<unsigned int>> parallelDijkstra(const std::vector<unsigned int>& src_ips, const std::set<unsigned int>& destinations) {
+        std::vector<std::vector<unsigned int>> results;
+
+        #pragma omp parallel for
+        for (unsigned long int i = 0; i < src_ips.size(); ++i) {
+            auto result = dijkstra(src_ips[i], destinations);
+            #pragma omp critical
+            {
+                results.emplace_back(std::move(result));
+            }
+        }
+        return results;
+    }
+    
+    void add_edge(const unsigned int &u, const unsigned int &v) {
+        graph[u].insert(v);
+        graph[v].insert(u);
+    }
+
+    std::vector<unsigned int> dijkstra(const unsigned int &start, const std::set<unsigned int> &destinations) {
+        if (destinations.find(start) != destinations.end()) {
+            return {start};
+        }
+
+        std::priority_queue<std::tuple<float, unsigned int>, std::vector<std::tuple<float, unsigned int>>, std::greater<std::tuple<float, unsigned int>>> min_heap;
+        std::unordered_map<unsigned int, float> distances;
```

Comment:
Can we change this to uint_fast8_t (https://en.cppreference.com/w/cpp/types/integer) to save memory? We don't need float and it's not gonna be more than 10 anyway.

Reply:
fixed, but I didn't find a noticeable improvement for the memory performance.